File: backend-api/app/schemas/feed.py
"""
Activity feed schemas for the Live Feed feature (F2.1-F2.3)
Hourly refresh, push notifications, relevance scoring
"""
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


class ActivityFeedItemSchema(BaseModel):
    """
    Single feed item in the activity stream (F2.1)
    Includes relevance scoring (F2.3) and local signal interpretation
    """
    id: int
    event_type: str = Field(..., description="funding, partnership, hiring, expansion, news")
    headline: str
    summary: Optional[str] = None

    # Company info
    company_name: Optional[str] = None
    company_domain: Optional[str] = None
    company_industry: Optional[str] = None

    # Location
    country: Optional[str] = None
    region: Optional[str] = None
    city: Optional[str] = None

    # Source
    source_url: Optional[str] = None
    source_name: Optional[str] = None

    # Relevance scoring (F2.3)
    relevance_score: float = Field(
        default=0.5,
        ge=0.0,
        le=1.0,
        description="AI-scored relevance 0.0-1.0 based on user's vertical"
    )
    is_high_priority: bool = Field(
        default=False,
        description="True if relevance_score >= 0.8 (buying signal)"
    )

    # Local signal interpretation (F2.3)
    local_context: Optional[str] = Field(
        None,
        description="Interpreted local language/context if applicable"
    )
    signal_type: Optional[str] = Field(
        None,
        description="growth, hiring, distress, expansion, partnership"
    )

    # Timestamps
    published_at: Optional[datetime] = None
    indexed_at: datetime

    class Config:
        from_attributes = True


class FeedResponse(BaseModel):
    """Paginated feed response with metadata"""
    items: List[ActivityFeedItemSchema]
    total_count: int
    page: int = 1
    page_size: int = 20
    has_more: bool = False
    last_updated: datetime = Field(default_factory=datetime.utcnow)
    # High priority items for push notifications (F2.2)
    high_priority_count: int = Field(
        default=0,
        description="Count of items with relevance_score >= 0.8"
    )


class FeedFilterOptions(BaseModel):
    """Available filter options for feed"""
    event_types: List[str] = ["funding", "partnership", "hiring", "expansion", "news"]
    countries: List[str] = ["Nigeria", "Ghana"]
    industries: List[str] = [
        "Fintech", "Banking", "Logistics", "E-commerce",
        "AgriTech", "HealthTech", "EdTech", "Other"
    ]


class PushNotificationPayload(BaseModel):
    """Push notification for high-relevance events (F2.2)"""
    feed_item_id: int
    headline: str
    company_name: Optional[str]
    event_type: str
    relevance_score: float
    action_url: str = Field(
        ...,
        description="Deep link to view full details"
    )

File: frontend-extension/src/shared/api/feed.ts
/**
 * Industry Feed API - News and updates for user's industry
 * Updated to match backend endpoint structure
 */
import { api } from "./client";
import type { IndustryNews, IndustryFeedResponse } from "./types";

// Legacy types for backward compatibility
export interface ActivityFeedItem {
  id: number;
  event_type: "funding" | "partnership" | "hiring" | "expansion" | "news";
  headline: string;
  summary?: string;
  company_name?: string;
  company_domain?: string;
  company_industry?: string;
  country?: string;
  region?: string;
  city?: string;
  source_url?: string;
  source_name?: string;
  relevance_score: number;
  is_high_priority: boolean;
  local_context?: string;
  signal_type?: string;
  published_at?: string;
  indexed_at: string;
}

export interface FeedResponse {
  items: ActivityFeedItem[];
  total_count: number;
  page: number;
  page_size: number;
  has_more: boolean;
  last_updated: string;
  high_priority_count: number;
}

export interface FeedStats {
  total_items: number;
  by_event_type: Record<string, number>;
  by_industry: Record<string, number>;
}

export interface EventType {
  value: string;
  label: string;
  description: string;
}

export const feedApi = {
  // Get industry-specific news feed based on organization's industry
  // Backend: GET /feed/industry
  getIndustryFeed: async (
    page = 1,
    limit = 20,
    newsType?: string
  ): Promise<IndustryFeedResponse> => {
    const params: Record<string, string | number> = {
      page,
      page_size: limit, // Backend uses page_size
    };
    if (newsType) params.news_type = newsType;

    const response = await api.get<IndustryFeedResponse>("/feed/industry", {
      params,
    });
    return response.data;
  },

  // Bookmark/unbookmark a news item
  // Backend: POST /feed/industry/bookmark with body { news_id: ... }
  toggleBookmark: async (newsId: string): Promise<{ bookmarked: boolean }> => {
    const response = await api.post<{ bookmarked: boolean }>(
      "/feed/industry/bookmark",
      {
        news_id: parseInt(newsId, 10),
      }
    );
    return response.data;
  },

  // Get bookmarked news
  // Backend: GET /feed/industry/bookmarks (returns array, no pagination)
  getBookmarked: async (): Promise<IndustryNews[]> => {
    const response = await api.get<IndustryNews[]>("/feed/industry/bookmarks");
    return response.data;
  },

  // Legacy feed endpoint (for backward compatibility)
  // Backend: GET /feed/
  getFeed: async (
    page = 1,
    pageSize = 20,
    eventType?: string,
    industry?: string
  ): Promise<FeedResponse> => {
    const params: Record<string, string | number> = { page, page_size: pageSize };
    if (eventType) params.event_type = eventType;
    if (industry) params.industry = industry;

    const response = await api.get<FeedResponse>("/feed/", { params });
    return response.data;
  },

  // Backend: GET /feed/refresh
  refreshFeed: async (): Promise<{
    status: string;
    items_added: number;
    timestamp: string;
  }> => {
    const response = await api.get("/feed/refresh");
    return response.data;
  },

  // Backend: GET /feed/stats
  getStats: async (): Promise<FeedStats> => {
    const response = await api.get<FeedStats>("/feed/stats");
    return response.data;
  },

  // Backend: GET /feed/event-types
  getEventTypes: async (): Promise<{ event_types: EventType[] }> => {
    const response = await api.get("/feed/event-types");
    return response.data;
  },

  // Get news types for industry feed
  // Backend: GET /feed/industry/news-types
  getNewsTypes: async (): Promise<{
    news_types: { value: string; label: string; description: string }[];
  }> => {
    const response = await api.get("/feed/industry/news-types");
    return response.data;
  },

  // Get available industries
  // Backend: GET /feed/industry/industries
  getIndustries: async (): Promise<{ industries: string[] }> => {
    const response = await api.get("/feed/industry/industries");
    return response.data;
  },
};
